### PR TITLE
Feature: allow public calendars

### PR DESCRIPTION
**Pull request by tchapi**
Override the default DAVACL plugin to allow public calendars, introducing a new `ACCESS_PUBLIC` constant

#### Public calendars

They are accessible without credentials, even in the browser

Screenshot 2023-08-15 at 21 22 37

#### Adds a new "Public" field when editing a calendar, to mark the calendar public (the user remains the sole owner as a public calendar cannot be shared for now)

Screenshot 2023-08-15 at 21 05 57

#### We add a badge in the list to indicate it's public

Screenshot 2023-08-15 at 21 06 08

Related PR: https://github.com/sabre-io/dav/pull/820

### CODE REVIEW

**Comment by 1Luc1** [comment]
Having events within the public calendar does trigger the authentication dialog.
grafik

I guess this entries doesn't have the `unauthenticated` principal
grafik

**Comment by 1Luc1** [comment]
Having a calender which is shared, can be changed to a public calender afterwards.
grafik

And the shared calender entry.
grafik

The Events got still synchronized. So it seems, this does work? Or should this be more restricted?

**Comment by david-raine** [comment]
Does this allow phonebooks to be shared?

**Comment by tchapi** [comment]
Thanks @1Luc1 for the tests!

> Having events (todos, notes also?) within the public calendar does trigger the authentication dialog.

That's weird, I didn't have the same behavior (with events only but having the other ones should not change much)

> Having a calender which is shared, can be changed to a public calender afterwards.

This is a bug, I'll fix asap

**Comment by tchapi** [comment]
Hello @david-raine 

> Does this allow phonebooks to be shared?

Unfortunately no, because the underlying sabre-io/dav library does not support it. See https://github.com/tchapi/davis/issues/87

File: src/Entity/CalendarInstance.php [hunk 1]
```diff
@@ -19,14 +19,29 @@
...
-    public const INVITE_STATUS_ACCEPTED = 2;
-    public const INVITE_STATUS_DECLINED = 3;
-    public const INVITE_STATUS_INVALID = 4;
+    public const INVITE_NORESPONSE = 1;
+    public const INVITE_ACCEPTED = 2;
+    public const INVITE_DECLINED = 3;
+    public const INVITE_INVALID = 4;
 
-    public const ACCESS_OWNER = 1;
+    public const ACCESS_NOTSHARED = 0;
```
Comment (tchapi) [inline src/Entity/CalendarInstance.php:27]: 🌱 This is mainly to comply with SabreDAV's namings

**Comment by 1Luc1** [comment]
Tested with branch [feature/public-calendars](https://github.com/tchapi/davis/tree/feature/public-calendars) 

> Thanks @1Luc1 for the tests!
> 
> > Having events (todos, notes also?) within the public calendar does trigger the authentication dialog.
> 
> That's weird, I didn't have the same behavior (with events only but having the other ones should not change much)

My I be more specific about the behavior. Calling the public calender with the link ( e.g.: davis.at/dav/calendars/public/test) within the browser (firefox) will trigger the authentication dialog only if there are events within the calender. If there are no events within the calender no authentication dialog will be called.

On the other hand, subscribing to the public calender within an email client (thunderbird) doesn't open any authentication dialog, even if there is a events within the calender.

So, this behavior is acceptable for my understanding. ✅

----

> > Having a calender which is shared, can be changed to a public calender afterwards.
> 
> ~This is a bug, I'll fix asap (_meaning: it should work, but it should also revoke all shared accesses since it's not used anymore and can lead to side effects_)~ ✅ Fixed

Tested ✅

**Comment by 1Luc1** [comment]
Found an interesting security issue, which is mainly a problem from the user and thunderbirds behavior.
If an user is created same name and the password is possible to add events to an public calendar within thunderbird. This will probably apply to non public calendars too.

Subscribe to a public calendar
grafik

Thunderbird shows the calendar as read only
grafik

Change read only within the calendar properties
grafik

Add events to calendar without providing a password :x:

Maybe add another security check within the user creation to not allow username and password the same?

**Comment by tchapi** [comment]
> will trigger the authentication dialog only if there are events within the calender. 

Oh ok, understood, thanks a lot for the clarification! I will test that behavior, I have a feeling it's some edge case on the sabre/dav side, but it should be fixable I guess, will work on it.

> If an user is created same name and the password is possible to add events to an public calendar

That's "funny" ... auth is managed by sabre/dav entirely, but it might be Thunderbird that tries to login optimistically by providing the username as a password? I'll try to reproduce

**Comment by tchapi** [comment]
> Calling the public calender with the link ( e.g.: davis.at/dav/calendars/public/test) within the browser (firefox) will trigger the authentication dialog only if there are events within the calender.

I think https://github.com/tchapi/davis/pull/105/commits/62d39b68893d1f2a97ed910ea0935f6a8e106539 should fix it

**Comment by 1Luc1** [comment]
> > Calling the public calender with the link ( e.g.: davis.at/dav/calendars/public/test) within the browser (firefox) will trigger the authentication dialog only if there are events within the calender.
> 
> I think [c1b0a00](https://github.com/tchapi/davis/commit/c1b0a008191af24f1de04b6902526d668e27f1e2) should fix it

Thanks. Just tested your changes and within the browser no Authentication Dialog appears. ✅

Unfortunately, within Thunderbird a  Authentication Dialog shows up subscribed to an public calender. ❌

**Comment by 1Luc1** [comment]
> > Unfortunately, within Thunderbird an authentication dialog shows up if subscribing to a public calender. ❌
> 
> This protocol is making me go nuts. Will have a look!

Can you tell, when you can get another look?

**Comment by tchapi** [comment]
> Can you tell, when you can get another look?

Unfortunately I'm quite busy on various other subjects so I can't really commit (as it takes quite some time to investigate the code for this feature); I'll do my best to finish next week

**Comment by 1Luc1** [comment]
Just checked your changes and tried to replicated the usecase where subscribing to a public calendar within thunderbird once worked, but couldn't really redo your changes to make it work. Now I doubt it ever worked :confounded:

**Comment by 1Luc1** [comment]
Did some further digin and found that within `getAcl` method from `PublicAwareDAVACLPlugin` adding the following checks

```
$node instanceof \Sabre\CalDAV\SharedCalendar
$node instanceof \Sabre\CalDAV\CalendarHome
$node instanceof \Sabre\CalDAV\Principal\User
```

within an own `if condition` returning `{DAV:}unauthenticated ...` won't show the authentication dialog within Thunderbird. So maybe that's something to work with?

**Comment by tchapi** [comment]
Thanks for your ideas. I've tried Thunderbird and could find a potential fix that I implemented in https://github.com/tchapi/davis/pull/105/commits/587493580a6d30d1cfc69e719476c45f9ab52297.

**Comment by tchapi** [comment]
Happy to get your feedback on it. Sorry this is taking quite some time 🙇🏼

**Comment by 1Luc1** [comment]
Thanks for your fix. Tested it as good as possible and subscribing without username/password works as intended. Unfortunately, it is now not possible as an owner of the calender to subscribe to the calender with username and password, within Thunderbird. This is of course necessary to write events into the "public" calendar.

**Comment by 1Luc1** [comment]
During testing I came across a use case, which should be covered within this PR. I mentioned it some [comments above](https://github.com/tchapi/davis/pull/105#issuecomment-1680009927) and thought it was a problem. Now I think the following use case should be covered:

Having a calender which should be shared with other users. The other users have read/write access to the shared calender, meaning a group of users can edit one calender. This calender can also be a public calender. 

A good example is having an event calender. Some people add/edit events within this calender and want to make the event calender public available to publicity.

Before you changed this behavior, you talked about [side effects](https://github.com/tchapi/davis/pull/105#issuecomment-1695223355). I'm curious, are there any?

**Comment by tchapi** [comment]
> Having a calender which should be shared with other users.

This will be another use case that I won't cover right now

> it is now not possible as an owner of the calender to subscribe to the calender with username and password, within Thunderbird.

I have tested and it is possible on my side. Either you provide the credentials and uncheck "read-only" — Thunderbird will ask for the password once (you can store it), and you can create event (as the owner of the calendar). Or you check "read only" and you don't need to authenticate, and will just retrieve the events

Below: private version (auth) is the calendar I used to create both events. They appear in the "public" version (of the same calendar)
Screenshot 2023-10-27 at 20 57 51

**Comment by tchapi** [comment]
So all in all, I think it's correctly working as of now (at last!). Will leave it open for a while and then merge to the master

**Comment by 1Luc1** [comment]
Thanks for re-testing. Your are right, it works as intended. Thanks again for your work!



>     Having a calender which should be shared with other users.
> 
> This will be another use case that I won't cover right now

Alright. Should a make a new Issue for this case?

**Comment by tchapi** [comment]
> Your are right, it works as intended. Thanks again for your work!

🎉 🎉 🎉 🎉 🎉 🎉 🎉 

> Should a make a new Issue for this case?

Yes please. Given the time it takes to fiddle with sabre/dav code, I'm not promising anything though sorry. It's really a pain to find circumvention ways for something that should be standard

**Comment by 1Luc1** [comment]
> > it is now not possible as an owner of the calender to subscribe to the calender with username and password, within Thunderbird.
> 
> I have tested and it is possible on my side. Either you provide the credentials and uncheck "read-only" — Thunderbird will ask for the password once (you can store it), and you can create event (as the owner of the calendar). Or you check "read only" and you don't need to authenticate, and will just retrieve the events

During testing for https://github.com/tchapi/davis/pull/200 i couldn't subscribe to a public calendar with username / password and add some events. Events won't show up. Tested with v5.1.1

@tchapi may you can test this case on your side?
